plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java

```diff
@@ -0,0 +1,303 @@
+            client.emit(event, runID.toString());
+        }
+        catch (IllegalAccessException | JsonProcessingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void queryCompleted(QueryCompletedEvent queryCompletedEvent)
+    {
+        UUID runID = getQueryId(queryCompletedEvent.getMetadata());
+
+        try {
+            OpenLineage.RunEvent event = getCompletedEvent(runID, queryCompletedEvent);
+            client.emit(event, runID.toString());
+        }
+        catch (IllegalAccessException | JsonProcessingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private Optional<OpenLineage.RunFacet> getTrinoQueryContextFacet(QueryContext queryContext)
+    {
+        if (this.trinoQueryContextFacetEnabled) {
+            OpenLineage.RunFacet queryContextFacet = ol.newRunFacet();
+
+            queryContextFacet
+                    .getAdditionalProperties()
+                    .put("serverVersion", queryContext.getServerVersion());
+            queryContextFacet
+                    .getAdditionalProperties()
+                    .put("environment", queryContext.getEnvironment());
+            queryContext.getQueryType().ifPresent(queryType ->
+                    queryContextFacet
+                            .getAdditionalProperties()
+                            .put("queryType", queryType));
+
+            return Optional.of(queryContextFacet);
+        }
+        return Optional.empty();
+    }
+
+    private Optional<OpenLineage.RunFacet> getTrinoMetadataFacet(QueryMetadata queryMetadata)
+    {
+        if (this.trinoMetadataFacetEnabled) {
+            OpenLineage.RunFacet trinoMetadataFacet = ol.newRunFacet();
+
+            queryMetadata.getPlan().ifPresent(
+                    queryPlan -> trinoMetadataFacet
+                            .getAdditionalProperties()
+                            .put("queryPlan", queryPlan));
+
+            queryMetadata.getTransactionId().ifPresent(
+                    transactionId -> trinoMetadataFacet
+                            .getAdditionalProperties()
+                            .put("transactionId", transactionId));
+
+            return Optional.of(trinoMetadataFacet);
+        }
+        return Optional.empty();
+    }
+
+    private Optional<OpenLineage.RunFacet> getTrinoQueryStatisticsFacet(QueryStatistics queryStatistics)
+            throws IllegalAccessException
+    {
+        if (this.queryStatisticsFacetEnabled) {
+            OpenLineage.RunFacet trinoQueryStatisticsFacet = ol.newRunFacet();
+
+            for (Field field : queryStatistics.getClass().getDeclaredFields()) {
+                field.setAccessible(true);
+                trinoQueryStatisticsFacet
+                        .getAdditionalProperties()
+                        .put(field.getName(), String.valueOf(field.get(queryStatistics)));
+            }
```

Comment:
I wonder if using here manual mapping of properties instead of reflection wouldn't be beneficiary in terms of stability and maintenance, currently we are not sure what should be expected in statistics facet

Reply:
I assumed additional facets do not need to have such strict contracts and this is much more convenient to maintain. Let's see if someone else has opinion in either direction.